subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/validation/MavenPublicationErrorChecker.java

```diff
@@ -43,14 +48,56 @@ public abstract class MavenPublicationErrorChecker extends PublicationErrorCheck
      * @throws PublishException if the artifacts are modified
      */
     public static void checkThatArtifactIsPublishedUnmodified(PublishArtifact source, DefaultMavenArtifactSet mainArtifacts) {
+        // Do we want to allow _additional_ artifacts to published? If not, we need to completely associate the source and artifacts
+        // at a higher level. As-is, this just verifies that none are removed from the components artifacts.
+        MavenArtifact bestMatch = null;
+        Set<Difference> bestMatchDifferences = null;
         for (MavenArtifact mavenArtifact : mainArtifacts) {
-            if (source.getFile().equals(mavenArtifact.getFile())
-                && source.getExtension().equals(mavenArtifact.getExtension())
-                && Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+            EnumSet<Difference> difference = EnumSet.noneOf(Difference.class);
+            if (!source.getFile().equals(mavenArtifact.getFile())) {
+                difference.add(Difference.FILE);
+            }
+            if (!source.getExtension().equals(mavenArtifact.getExtension())) {
+                difference.add(Difference.EXTENSION);
+            }
+            if (!Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+                difference.add(Difference.CLASSIFIER);
+            }
+            // If it's all equal, we found a matching artifact that is being published
+            if (difference.isEmpty()) {
                 return;
             }
+            if (bestMatch == null || difference.size() < bestMatchDifferences.size()) {
+                bestMatch = mavenArtifact;
+                bestMatchDifferences = difference;
+            }
         }
+        if (bestMatch == null) {
+            throw new PublishException("Cannot publish module metadata because it does not contain any artifacts.");
+        }
+        String differences = getDifferences(source, bestMatch, bestMatchDifferences);
+        throw new PublishException("Cannot publish module metadata because an artifact from a component has been removed. The best match had these problems:\n" + differences);
+    }
+
+    private static String getDifferences(PublishArtifact source, MavenArtifact bestMatch, Set<Difference> bestMatchDifferences) {
+        return bestMatchDifferences.stream().map(diff -> {
```

Comment:
❗ This is kind of odd because all 3 tests here (file, extension and classifier could fail) could fail together.  And if extension fails, then file MUST also fail.  Yet the error message only would mention one thing at a time.

So if File and Classifier both differ, you would see a File failure message.  Then, upon correcting that, you would see a Classifier failure message.  This isn't a great UX - could all the messages be shown initially?

Reply:
I'm not sure what you mean -- all of the failures are shown, that's the point of using `reduce` to add them all together.

Reply:
My mistake, I misread this.